### DISCUSSION TITLE
Vertically align text in a Rect when using insert_textbox

### DISCUSSION THREAD

**Discussion by rafayaar** [pymupdf/PyMuPDF#3993] [General]
Hi @JorjMcKie 
I have 2 things to ask

- I am using `insert_textbox` to insert text within provided `rect`. I can easily horizontal align using `align`, but there is no direct way to vertically align.
When text is insert in a rect, it is insert on top left. I would want it to vertically align.
How can I vertically align a text in Rect.
This discussion didn't really helped me out cause it uses textwriter in text [Link-Discussion-1662](https://github.com/pymupdf/PyMuPDF/discussions/1662#discussioncomment-2503486)

- I am also rotating the text within text box. I am using `morph=(point,matrix)`. Origin of rotation is top left as we have point system with us. I can calculate width/2 and height/2 of Rect and rotate accordingly but text rotates based on center of Rect and NOT OF TEXT actually. 

Would really appreciate your answer. Thanks

**Comment by rafayaar** [comment]
![image](https://github.com/user-attachments/assets/d7b37e99-4c92-47ce-98b9-016763124011)

I am trying to re render attached PDF. Attached screenshot is where I am having issues

[dinnerly 2.pdf](https://github.com/user-attachments/files/17541346/dinnerly.2.pdf)

**Comment by JorjMcKie** [comment]
# First Point
There indeed is no direct support of centering (aligning???) vertically. But with a small effort you can achieve the effect.
Just execute on a temporary page and check the return code (`rc`): this is the height of the unused sub-rectangle at the bottom.
Use the half of this value to write to a accordingly defined other rectangle `page.insert_textbox(rect + (0, rc/2, 0, rc/2), ...)`.
This modified rect is shifted downwards.
To compute `rc`:
```python
temp = pymupdf.open()
tpage=temp.new_page(width=rect.width, height=rect.height)
rc = tpage.insert_textbox(rect, ...)
temp.close()
```

**Reply by rafayaar** [reply]
This itself adjust the rect but not the text inside it. My goal is to vertically center text in a Rect

**Reply by JorjMcKie** [reply]
But the effect looks like vertically centered text in the original rect:
```python
import pymupdf

text = "show vertically centered in the original rectangle"
rect = pymupdf.Rect(100, 100, 200, 200)
doc = pymupdf.open()
page = doc.new_page()
temp = pymupdf.open()
tpage = temp.new_page(width=rect.width, height=rect.height)
rc = tpage.insert_textbox(rect, text)
page.insert_textbox(rect + (0, rc / 2, 0, rc / 2), text)
page.draw_rect(rect, color=(1, 0, 0))
doc.ez_save(__file__.replace(".py", ".pdf"))
```
![image](https://github.com/user-attachments/assets/8c7d2c86-6ef4-4b9d-a9d2-64428b511f5b)
What are your objections?

**Reply by rafayaar** [reply]
Yes this helps.
My goal is to vertically align the text and from the center of the text itself do a rotation.
Currently if I apply morph it rotates from top left of the text. I want rotation from center of text such that it rotates from center origin

**Reply by rafayaar** [reply]
Increase the font size and the text will be pushed towards bottom

**Comment by JorjMcKie** [comment]
As per your other point:
* We cannot write text along some curve - only straight lines. This will not be supported ever presumably.
* As far as I can see, the only way to approximately achieve what you want is
   * make an upright version,
   * then make an image of the result
   * insert this image as a temporary PDF page via `page.show_pdf_page(... ,rotate=theta)` with a desired angle.

**Comment by rafayaar** [comment]
@JorjMcKie I have observed that `insert_textbox` comes with padding. Please correct me if I am wrong

**Reply by JorjMcKie** [reply]
What does padding mean?

**Reply by rafayaar** [reply]
When text is inserted into textbox it keeps some space inside the borders of a rect. Even though there is visibly some space left to fill entire textbox, it just doesn't render it at all.
I extract a PDF using get_text block methods and for each text element its rect. When i redraw using same dimensions and same font family,size it doesn't render the text in textbox. Text appear when either I reduce the font size a bit `(e.g) *0.9` or increase rect dimesnions
